### DISCUSSION TITLE
Mechanism for executing code before running notebook?

### DISCUSSION THREAD

**Discussion by matthew-brett** [executablebooks/meta#172] [Q&A]
In Bookdown, you can specify a [before_chapter_script](https://bookdown.org/yihui/bookdown/configuration.html) to run code before the notebook (chapter) runs.   The typical use-case is setting random number seeds, so the rendered output us predictable, but the notebook itself does not get polluted with seeding code.

I was doing this via some scripting in my own hacked up version of the previous iteration of the Jupyter Book - but is there a way to do this in the current version?

**Comment by choldgraf** [comment]
Right now it's not possible, though something that has been discussed.

I think we're tracking the idea here: https://github.com/executablebooks/jupyter-book/issues/777
and sorta related here: https://github.com/executablebooks/MyST-NB/issues/274

The main reason this hasn't been implemented yet is because thus far we've had a fairly strong 1-to-1 mapping of "content file" and "page in the book" to make the contents more self-contained and reproducible. Though this is a feature many have requested.

**Reply by matthew-brett** [reply]
In teaching, its very common to have notebooks with random content, where some random sequences would be very unhelpful.  That's very uncomfortable, because it means that each time we render the content, we take the risk that the page will become confusing, perhaps very confusing.   Or we have to put stuff into the notebook that really shouldn't appear in the version the student runs, with random seeding.

Other evidence that this is an important feature:

* Its presence in bookdown
* It's fairly often used in Bookdown - from a quick search on github for "before_chapter_script"
* Hadley Wickham uses it in r4ds - exectly to set a  random seed : https://github.com/hadley/r4ds/blob/master/_common.R#L1
* I am using it in my Bookdown project for the same reason : https://github.com/resampling-stats/resampling-with/blob/master/source/_common.R
* I needed it enough in JupyterBook that I wrote code to insert this header stuff in every notebook before rendering : https://github.com/matthew-brett/dsfe2019/blob/master/scripts/generate_textbook.py#L134

**Reply by choldgraf** [reply]
well it's possible to have the content in jupyter book, just put a `remove-cell` tag on any code that you want removed from the "book" version. Then it will remain in the *source*, just not in the rendered book.

The reason that *injecting* code hasn't happened is because then it will seem like there *are* things in the book version that are *not* in the source. E.g., any Binder links wouldn't work for the book.

**Reply by matthew-brett** [reply]
I may have misunderstood, but what I need in this case is for the code to run only in the case of the rendered notebook, and for the code not to appear in the rendered version *or the source*, that the student gets by downloading or running the notebook in JupyterHub.

But in general - isn't this a use-at-your-own-risk type of situation?  Where we the authors have to be careful we don't put pre-injected code in, that breaks the downloaded notebook, when missing.

**Reply by choldgraf** [reply]
I'm not personally opposed to the feature, as you say I think it's common across many use-cases. I think many folks want to use this to do things like standardized imports across all of their pages, which _would_ break things. Others have wanted to use it for standardizing things like matplotlib formatting across pages. What I'd personally worry about is that one feature request would lead to another, and if we let people inject code on each page, in a few weeks we'd start seeing issues about why the notebook code and their book outputs don't look the same etc. AKA, "use at your own risk" is reasonable IMO but also might generate maintainer burden and confusion for people regardless.

**Reply by akhmerov** [reply]
If I were to implement this in my own code, I would do this by using an external script that prepends initialization code to each document. Since myst-nb executes the notebooks extremely early in the pipeline (https://github.com/executablebooks/meta/discussions/144), this would need to happen outside of sphinx.

**Reply by matthew-brett** [reply]
@choldgraf - honestly I think you do not give users enough credit here.  If there is a warning all over this feature, with explanations, I bet you'll find it is used carefully.

**Reply by choldgraf** [reply]
again I'm not against the feature! Just surfacing points to discuss. I think the place to give feedback like the helpful references to other implementations in R above would be in https://github.com/executablebooks/jupyter-book/issues/777. If you're interested in trying out a PR to offer up a prototype, I think that the code injection would need to be somewhere before here: https://github.com/executablebooks/MyST-NB/blob/master/myst_nb/parser.py#L74
